### PR TITLE
[IE CLDNN] Implement ExtractImagePatches operation

### PR DESCRIPTION
**Pull request by echuraev**
The ExtractImagePatches operation collects patches from the input
tensor, as if applying a convolution. All extracted patches are stacked
in the depth dimension of the output.

JIRA: 30055

### CODE REVIEW

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/extract_image_patches/extract_image_patches_kernel_base.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        MakeJitConstant("SIZE_ROWS", params.sizes[0]),
+        MakeJitConstant("SIZE_COLS", params.sizes[1]),
+        MakeJitConstant("STRIDE_ROWS", params.strides[0]),
+        MakeJitConstant("STRIDE_COLS", params.strides[1]),
+        MakeJitConstant("RATES_ROWS", params.rates[0]),
+        MakeJitConstant("RATES_COLS", params.rates[1]),
+    });
+    if (params.auto_pad == "same_upper")
+        jit.AddConstant(MakeJitConstant("AUTO_PAD", 1));
+    if (params.auto_pad == "same_lower")
```
Comment (sshlyapn) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/extract_image_patches/extract_image_patches_kernel_base.cpp:49]: `else if` will be better here

File: inference-engine/thirdparty/clDNN/tests/test_cases/extract_image_patches_gpu_test.cpp [hunk 2]
```diff
@@ -0,0 +1,595 @@
...
+using namespace cldnn;
+using namespace tests;
+
+TEST(extract_image_patches_gpu, basic) {
+    //  Input  : 1x1x10x10
+    //  Output : 1x9x2x2
+
+    tensor output_shape = {1, 9, 2, 2};
+    //const auto& engine = get_test_engine();
+    engine engine(engine_configuration(false, false, false, "", "", true, "", "/tmp/sources/"));
```
Comment (sshlyapn) [inline inference-engine/thirdparty/clDNN/tests/test_cases/extract_image_patches_gpu_test.cpp:38]: Debug code

File: inference-engine/thirdparty/clDNN/tests/test_cases/extract_image_patches_gpu_test.cpp [hunk 3]
```diff
@@ -0,0 +1,595 @@
...
+    //}
+    //std::cout << "\nOUTPUT!" << std::endl;
+    //std::cout << "ANSWERS:" << std::endl;
+    //for (size_t i = 0; i < answers.size(); ++i) {
+    //    if (i % 2)
+    //        std::cout << " " << answers[i];
+    //    else
+    //        std::cout << "\n" << answers[i];
+    //}
+    //std::cout << "\nANSWERS!" << std::endl;
```
Comment (sshlyapn) [inline inference-engine/thirdparty/clDNN/tests/test_cases/extract_image_patches_gpu_test.cpp:116]: Debug code

File: inference-engine/thirdparty/clDNN/tests/test_cases/extract_image_patches_gpu_test.cpp [hunk 4]
```diff
@@ -0,0 +1,595 @@
...
+    std::vector<unsigned int> rates = {1, 1};
+    std::string auto_pad = "valid";
+
+    auto input = memory::allocate(engine, { data_types::f32, format::bfyx, { batch, depth, in_rows, in_cols } });
+
+    std::vector<float> inputVals(batch * depth * in_rows * in_cols);
+    std::generate(inputVals.begin(), inputVals.end(), []() {
+                static float n = 1;
+                return n++;
+            });
```
Comment (sshlyapn) [inline inference-engine/thirdparty/clDNN/tests/test_cases/extract_image_patches_gpu_test.cpp:54]: I think this codestyle will be better:
```
    std::generate(inputVals.begin(), inputVals.end(), []() {
        static float n = 1;
        n++;
    });
```

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/extract_image_patches_ref.cl [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    const uint col = cur_col_ind +
+               STRIDE_COLS * out_col - col_padding +
+               (RATES_COLS - 1) * cur_col_ind;
+
+    const uint depth = out_depth % INPUT0_FEATURE_NUM;
+    const uint in_ind = INPUT0_GET_INDEX_SAFE(batch, depth, row, col);
+    const uint out_ind = OUTPUT_GET_INDEX(batch, out_depth, out_row, out_col);
+    OUTPUT_TYPE res = TO_OUTPUT_TYPE(input[in_ind]);
+    if (row < 0 || col < 0 || row >= INPUT0_SIZE_Y || col >= INPUT0_SIZE_X)
+        res = OUTPUT_VAL_ZERO;
```
Comment (sshlyapn) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/extract_image_patches_ref.cl:60]: Looks like this condition is needed only if AUTO_PAD macro defined? Am I right?

File: inference-engine/thirdparty/clDNN/tests/test_cases/extract_image_patches_gpu_test.cpp [hunk 6]
```diff
@@ -0,0 +1,577 @@
...
+    auto batch = 1;
+    auto depth = 1;
+    auto in_rows = 10;
+    auto in_cols = 10;
+    std::vector<unsigned int> sizes = {3, 3};
+    std::vector<unsigned int> strides = {5, 5};
+    std::vector<unsigned int> rates = {1, 1};
+    std::string auto_pad = "valid";
+
+    auto input = memory::allocate(engine, { data_types::f32, format::bfyx, { batch, depth, in_rows, in_cols } });
```
Comment (sshlyapn) [inline inference-engine/thirdparty/clDNN/tests/test_cases/extract_image_patches_gpu_test.cpp:47]: Note that tensor constructor is `tensor(value_type batch_num, value_type feature_num, value_type width, value_type height)`, so order `batch, depth, in_cols, in_rows` would be more appropriate

Comment (echuraev) [reply inference-engine/thirdparty/clDNN/tests/test_cases/extract_image_patches_gpu_test.cpp:47]: Thank you. I've changed in other places and forgot here. I'll update.

**Pull request merged**